kadi/commands/observations.py

```diff
@@ -181,16 +182,76 @@ def convert_aostrcat_to_acatable(obs, params):
     return aca
 
 
-def get_starcats(obsid=None, *, start=None, stop=None, set_ids=True, scenario=None, cmds=None):
-    """Get star catalogs corresponding to input parameters.
+def get_starcats_as_table(start=None, stop=None, *, obsid=None, unique=False,
+                          scenario=None, cmds=None):
+    """Get a single table of star catalog entries corresponding to input parameters.
 
-    The ``obsid``, ``start``, and ``stop`` parameters serve as matching filters
+    This function calls ``get_starcats`` with the same parameters and then
+    concatenates the results into a single table for convenience. In addition
+    to the usual star catalog columns, the ``obsid`` and ``starcat_date`` are
+    included.
+
+    The ``unique`` parameter can be set to ``True`` to only return unique star
+    catalog entries. There are numerous instances of a single commanded star
+    catalogs that is associated with two ObsIDs, for instance ACIS undercover
+    observations. To get only the first one, set ``unique=True``.
```

Comment:
For the unique flag, the first undercover I tried did not line up with expectations:

```
In [29]: get_starcats_as_table(start='2020:366:17:04:20.000', stop='2020:366:23:00:00.000')
Out[29]: 
<Table length=22>
obsid      starcat_date      slot  idx      id    type  sz         mag          maxmag          yang                zang         dim   res  halfw
int64         str21         int64 int64   int64   str3 str3      float64       float64        float64             float64       int64 int64 int64
----- --------------------- ----- ----- --------- ---- ---- ----------------- --------- ------------------- ------------------- ----- ----- -----
24882 2020:366:17:04:20.000     0     1         1  FID  8x8               7.0       8.0 -1174.9607093061636  -468.1481719926585     1     1    25
24882 2020:366:17:04:20.000     1     2         3  FID  8x8               7.0       8.0 -1177.6232024219992   561.4018012757298     1     1    25
24882 2020:366:17:04:20.000     2     3         4  FID  8x8               7.0       8.0  1223.4393109616512   563.8018390366224     1     1    25
24882 2020:366:17:04:20.000     3     4 552075520  BOT  6x6 8.885130882263184 10.390625  -912.6106246537174  -703.8856678520224     8     1    60
24882 2020:366:17:04:20.000     4     5 552086952  BOT  6x6 8.897590637207031  10.40625 -1510.7731283644346  -2392.854408253782     8     1    60
24882 2020:366:17:04:20.000     5     6 552206400  BOT  6x6 8.648664474487305  10.15625  -84.90442138289295  -2367.635642290802    12     1    80
24882 2020:366:17:04:20.000     6     7 552206936  BOT  6x6 6.875301837921143     8.375  2104.7955129097486  1092.1642810692815    16     1   100
24882 2020:366:17:04:20.000     7     8 552213664  BOT  6x6  8.73717975616455 10.234375    1073.65803292352  -944.4668937469213    12     1    80
24882 2020:366:17:04:20.000     0     9 552217088  ACQ  6x6 8.083139419555664   9.59375  168.79558020035054 -2223.1544736547244    12     1    80
24882 2020:366:17:04:20.000     1    10 552081976  ACQ  6x6  9.25229549407959     10.75 -1094.5981762913334  1597.2393211928293     8     1    60
24882 2020:366:17:04:20.000     2    11 552084112  ACQ  6x6  9.63385009765625 11.140625  -782.1356873895911  -292.7544209467494     8     1    60
24882 2020:366:17:04:20.000     0     1         1  FID  8x8               7.0       8.0 -1174.9607093061636  -468.1481719926585     1     1    25
24882 2020:366:17:04:20.000     1     2         3  FID  8x8               7.0       8.0 -1177.6232024219992   561.4018012757298     1     1    25
24882 2020:366:17:04:20.000     2     3         4  FID  8x8               7.0       8.0  1223.4393109616512   563.8018390366224     1     1    25
24882 2020:366:17:04:20.000     3     4 552075520  BOT  6x6 8.885130882263184 10.390625  -912.6106246537174  -703.8856678520224     8     1    60
24882 2020:366:17:04:20.000     4     5 552086952  BOT  6x6 8.897590637207031  10.40625 -1510.7731283644346  -2392.854408253782     8     1    60
24882 2020:366:17:04:20.000     5     6 552206400  BOT  6x6 8.648664474487305  10.15625  -84.90442138289295  -2367.635642290802    12     1    80
24882 2020:366:17:04:20.000     6     7 552206936  BOT  6x6 6.875301837921143     8.375  2104.7955129097486  1092.1642810692815    16     1   100
24882 2020:366:17:04:20.000     7     8 552213664  BOT  6x6  8.73717975616455 10.234375    1073.65803292352  -944.4668937469213    12     1    80
24882 2020:366:17:04:20.000     0     9 552217088  ACQ  6x6 8.083139419555664   9.59375  168.79558020035054 -2223.1544736547244    12     1    80
24882 2020:366:17:04:20.000     1    10 552081976  ACQ  6x6  9.25229549407959     10.75 -1094.5981762913334  1597.2393211928293     8     1    60
24882 2020:366:17:04:20.000     2    11 552084112  ACQ  6x6  9.63385009765625 11.140625  -782.1356873895911  -292.7544209467494     8     1    60
```

```

In [30]: get_starcats_as_table(start='2020:366:17:04:20.000', stop='2020:366:23:00:00.000', unique=True)
Out[30]: 
<Table length=11>
obsid      starcat_date      slot  idx      id    type  sz         mag          maxmag          yang                zang         dim   res  halfw
int64         str21         int64 int64   int64   str3 str3      float64       float64        float64             float64       int64 int64 int64
----- --------------------- ----- ----- --------- ---- ---- ----------------- --------- ------------------- ------------------- ----- ----- -----
24882 2020:366:17:04:20.000     0     1         1  FID  8x8               7.0       8.0 -1174.9607093061636  -468.1481719926585     1     1    25
24882 2020:366:17:04:20.000     1     2         3  FID  8x8               7.0       8.0 -1177.6232024219992   561.4018012757298     1     1    25
24882 2020:366:17:04:20.000     2     3         4  FID  8x8               7.0       8.0  1223.4393109616512   563.8018390366224     1     1    25
24882 2020:366:17:04:20.000     3     4 552075520  BOT  6x6 8.885130882263184 10.390625  -912.6106246537174  -703.8856678520224     8     1    60
24882 2020:366:17:04:20.000     4     5 552086952  BOT  6x6 8.897590637207031  10.40625 -1510.7731283644346  -2392.854408253782     8     1    60
24882 2020:366:17:04:20.000     5     6 552206400  BOT  6x6 8.648664474487305  10.15625  -84.90442138289295  -2367.635642290802    12     1    80
24882 2020:366:17:04:20.000     6     7 552206936  BOT  6x6 6.875301837921143     8.375  2104.7955129097486  1092.1642810692815    16     1   100
24882 2020:366:17:04:20.000     7     8 552213664  BOT  6x6  8.73717975616455 10.234375    1073.65803292352  -944.4668937469213    12     1    80
24882 2020:366:17:04:20.000     0     9 552217088  ACQ  6x6 8.083139419555664   9.59375  168.79558020035054 -2223.1544736547244    12     1    80
24882 2020:366:17:04:20.000     1    10 552081976  ACQ  6x6  9.25229549407959     10.75 -1094.5981762913334  1597.2393211928293     8     1    60
24882 2020:366:17:04:20.000     2    11 552084112  ACQ  6x6  9.63385009765625 11.140625  -782.1356873895911  -292.7544209467494     8     1    60
```

Yes, unique reduced this to one set of entries.  I just thought the second set of entries would have obsid 62647 in the table.

Reply:
This looks like something else that is behaving a little bit oddly with caching (and now my cache has a different state / obsid.

```
In [10]: with conf.set_temp('cache_starcats', True):
    ...:     print(get_starcats_as_table(start='2020:366:17:04:20.000', stop='2020:366:23:00:00.000'))
    ...: 
obsid      starcat_date     slot idx     id    type  sz        mag          maxmag          yang                zang        dim res halfw
----- --------------------- ---- --- --------- ---- --- ----------------- --------- ------------------- ------------------- --- --- -----
62647 2020:366:17:04:20.000    0   1         1  FID 8x8               7.0       8.0 -1174.9607093061636  -468.1481719926585   1   1    25
62647 2020:366:17:04:20.000    1   2         3  FID 8x8               7.0       8.0 -1177.6232024219992   561.4018012757298   1   1    25
62647 2020:366:17:04:20.000    2   3         4  FID 8x8               7.0       8.0  1223.4393109616512   563.8018390366224   1   1    25
62647 2020:366:17:04:20.000    3   4 552075520  BOT 6x6 8.885130882263184 10.390625  -912.6106246537174  -703.8856678520224   8   1    60
62647 2020:366:17:04:20.000    4   5 552086952  BOT 6x6 8.897590637207031  10.40625 -1510.7731283644346  -2392.854408253782   8   1    60
62647 2020:366:17:04:20.000    5   6 552206400  BOT 6x6 8.648664474487305  10.15625  -84.90442138289295  -2367.635642290802  12   1    80
62647 2020:366:17:04:20.000    6   7 552206936  BOT 6x6 6.875301837921143     8.375  2104.7955129097486  1092.1642810692815  16   1   100
62647 2020:366:17:04:20.000    7   8 552213664  BOT 6x6  8.73717975616455 10.234375    1073.65803292352  -944.4668937469213  12   1    80
62647 2020:366:17:04:20.000    0   9 552217088  ACQ 6x6 8.083139419555664   9.59375  168.79558020035054 -2223.1544736547244  12   1    80
62647 2020:366:17:04:20.000    1  10 552081976  ACQ 6x6  9.25229549407959     10.75 -1094.5981762913334  1597.2393211928293   8   1    60
62647 2020:366:17:04:20.000    2  11 552084112  ACQ 6x6  9.63385009765625 11.140625  -782.1356873895911  -292.7544209467494   8   1    60
62647 2020:366:17:04:20.000    0   1         1  FID 8x8               7.0       8.0 -1174.9607093061636  -468.1481719926585   1   1    25
62647 2020:366:17:04:20.000    1   2         3  FID 8x8               7.0       8.0 -1177.6232024219992   561.4018012757298   1   1    25
62647 2020:366:17:04:20.000    2   3         4  FID 8x8               7.0       8.0  1223.4393109616512   563.8018390366224   1   1    25
62647 2020:366:17:04:20.000    3   4 552075520  BOT 6x6 8.885130882263184 10.390625  -912.6106246537174  -703.8856678520224   8   1    60
62647 2020:366:17:04:20.000    4   5 552086952  BOT 6x6 8.897590637207031  10.40625 -1510.7731283644346  -2392.854408253782   8   1    60
62647 2020:366:17:04:20.000    5   6 552206400  BOT 6x6 8.648664474487305  10.15625  -84.90442138289295  -2367.635642290802  12   1    80
62647 2020:366:17:04:20.000    6   7 552206936  BOT 6x6 6.875301837921143     8.375  2104.7955129097486  1092.1642810692815  16   1   100
62647 2020:366:17:04:20.000    7   8 552213664  BOT 6x6  8.73717975616455 10.234375    1073.65803292352  -944.4668937469213  12   1    80
62647 2020:366:17:04:20.000    0   9 552217088  ACQ 6x6 8.083139419555664   9.59375  168.79558020035054 -2223.1544736547244  12   1    80
62647 2020:366:17:04:20.000    1  10 552081976  ACQ 6x6  9.25229549407959     10.75 -1094.5981762913334  1597.2393211928293   8   1    60
62647 2020:366:17:04:20.000    2  11 552084112  ACQ 6x6  9.63385009765625 11.140625  -782.1356873895911  -292.7544209467494   8   1    60

In [11]: with conf.set_temp('cache_starcats', False):
    ...:     print(get_starcats_as_table(start='2020:366:17:04:20.000', stop='2020:366:23:00:00.000'))
    ...: 
obsid      starcat_date     slot idx     id    type  sz        mag          maxmag          yang                zang        dim res halfw
----- --------------------- ---- --- --------- ---- --- ----------------- --------- ------------------- ------------------- --- --- -----
24882 2020:366:17:04:20.000    0   1         1  FID 8x8               7.0       8.0 -1174.9607093061636  -468.1481719926585   1   1    25
24882 2020:366:17:04:20.000    1   2         3  FID 8x8               7.0       8.0 -1177.6232024219992   561.4018012757298   1   1    25
24882 2020:366:17:04:20.000    2   3         4  FID 8x8               7.0       8.0  1223.4393109616512   563.8018390366224   1   1    25
24882 2020:366:17:04:20.000    3   4 552075520  BOT 6x6 8.885130882263184 10.390625  -912.6106246537174  -703.8856678520224   8   1    60
24882 2020:366:17:04:20.000    4   5 552086952  BOT 6x6 8.897590637207031  10.40625 -1510.7731283644346  -2392.854408253782   8   1    60
24882 2020:366:17:04:20.000    5   6 552206400  BOT 6x6 8.648664474487305  10.15625  -84.90442138289295  -2367.635642290802  12   1    80
24882 2020:366:17:04:20.000    6   7 552206936  BOT 6x6 6.875301837921143     8.375  2104.7955129097486  1092.1642810692815  16   1   100
24882 2020:366:17:04:20.000    7   8 552213664  BOT 6x6  8.73717975616455 10.234375    1073.65803292352  -944.4668937469213  12   1    80
24882 2020:366:17:04:20.000    0   9 552217088  ACQ 6x6 8.083139419555664   9.59375  168.79558020035054 -2223.1544736547244  12   1    80
24882 2020:366:17:04:20.000    1  10 552081976  ACQ 6x6  9.25229549407959     10.75 -1094.5981762913334  1597.2393211928293   8   1    60
24882 2020:366:17:04:20.000    2  11 552084112  ACQ 6x6  9.63385009765625 11.140625  -782.1356873895911  -292.7544209467494   8   1    60
24882 2020:366:17:04:20.000    0   1         1  FID 8x8               7.0       8.0 -1174.9607093061636  -468.1481719926585   1   1    25
24882 2020:366:17:04:20.000    1   2         3  FID 8x8               7.0       8.0 -1177.6232024219992   561.4018012757298   1   1    25
24882 2020:366:17:04:20.000    2   3         4  FID 8x8               7.0       8.0  1223.4393109616512   563.8018390366224   1   1    25
24882 2020:366:17:04:20.000    3   4 552075520  BOT 6x6 8.885130882263184 10.390625  -912.6106246537174  -703.8856678520224   8   1    60
24882 2020:366:17:04:20.000    4   5 552086952  BOT 6x6 8.897590637207031  10.40625 -1510.7731283644346  -2392.854408253782   8   1    60
24882 2020:366:17:04:20.000    5   6 552206400  BOT 6x6 8.648664474487305  10.15625  -84.90442138289295  -2367.635642290802  12   1    80
24882 2020:366:17:04:20.000    6   7 552206936  BOT 6x6 6.875301837921143     8.375  2104.7955129097486  1092.1642810692815  16   1   100
24882 2020:366:17:04:20.000    7   8 552213664  BOT 6x6  8.73717975616455 10.234375    1073.65803292352  -944.4668937469213  12   1    80
24882 2020:366:17:04:20.000    0   9 552217088  ACQ 6x6 8.083139419555664   9.59375  168.79558020035054 -2223.1544736547244  12   1    80
24882 2020:366:17:04:20.000    1  10 552081976  ACQ 6x6  9.25229549407959     10.75 -1094.5981762913334  1597.2393211928293   8   1    60
24882 2020:366:17:04:20.000    2  11 552084112  ACQ 6x6  9.63385009765625 11.140625  -782.1356873895911  -292.7544209467494   8   1    60
```

Reply:
So if the starcats are just in there with one key, I would assume that get_starcats or get_starcats_as_table would need to hold on to the obsid to assign it in a way that makes sense for the undercover.  This is not new for this PR but I think is something I missed in the PR #226 review.

Reply:
Awesome testing @jeanconn thanks!! 

I think I just missed putting the `obsid` in the cache key. I'm actually not sure the `source` (load name) really matters but it doesn't hurt. See c4775e8. Does this key seem unique to you now?

Reply:
Thanks!  I wasn't sure that we actually needed to put the catalog in the cache twice for these observations that share catalog?  But using a unique keys does seem simpler than pulling out the one star catalog twice and labeling it again with `obs['obsid']` as part of the de-caching in get_starcats.